### PR TITLE
[JIMT][CT-446] upload files to codebridge

### PR DESCRIPTION
**Pull request by thomasoniii**
Adds a file upload option. Button is visible both on the root from the file uploader, as well as in the pop up from the kebab menu on folders.

https://github.com/user-attachments/assets/ad283f53-73b6-4a91-b97d-2202452d7bce

Things of note:
- This also adds an image viewer option into `Editor`, which uses a `visibileImageFileType` function to display a simple`img` tag. (though pythonlab here currently only allows `text/*` mimetypes)
- It is smart enough to prevent uploading files w/invalid names. It also patches up a few bugs in what names were actually considered valid.
- It's also smart enough to prevent uploading of duplicate folders/files
- Does some of the initial work for CT-822, because I was adding new hooks and just added them into separate files instead of growing the `FileBrowser` _even more_.
- #61253 was supposed to include support to prevent dragging files/folders into locations that would have duplicate names, but I must have lost a commit on that PR so the functionality was never released. Adds it into this PR instead.
- The `<FileUploader>` component which does the uploading is super generic, and we may want to put it into a different location. It's just a wrapper that provides a click handler (which is intercepted and _not_ propagated) around whatever content you want, and a callback for error states as well as file contents.
- Uses the same list of mime-types added in #58771 to prevent invalid file uploads.
- We had two names for the same action - `New File` in the root, and `Add file` into a sub folder. So it could be confusing about them doing different things - especially since this PR would now add an `Upload file` option right next to `Add file`. So this PR drops the `Add file` text in favor of `New file`.
- Also changes the other root level action from `New Folder` -> `New folder` for consistency.
- Includes analytics - will fire when any file is uploaded successfully, when a file is NOT uploaded successfully (reader error - this is unlikely), when a file is uploaded with an invalid name, and when a file is uploaded which is not allowed (improper file type)

## Links

- jira ticket: [CT-446](https://codedotorg.atlassian.net/browse/CT-446)

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: apps/src/codebridge/Editor/index.tsx [hunk 1]
```diff
@@ -38,6 +38,15 @@ export const Editor = ({langMapping, editableFileTypes}: EditorProps) => {
...
 
+  if (file && viewableImageFileType(file.language)) {
+    const base64 = window.btoa(file.contents);
+    return (
+      <div>
+        <img src={`data:image/png;base64,${base64}`} alt={file.name} />
+      </div>
+    );
+  }
+
```
Comment (thomasoniii) [inline apps/src/codebridge/Editor/index.tsx:49]: simple viewer - if an image file is selected (gif, jpg, png), then just display an image.

File: apps/src/codebridge/FileBrowser/FileUploader.tsx [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+        };
+      }
+    }, [callback, errorCallback]);
+
+    return (
+      <div
+        onClick={e => {
+          e.stopPropagation();
+
+          inputRef.current?.click();
```
Comment (thomasoniii) [inline apps/src/codebridge/FileBrowser/FileUploader.tsx:95]: This prevents the popup from closing, and then just invisibly clicks on the invisible file upload input.

File: apps/src/codebridge/FileBrowser/hooks/useCheckForDuplicateFilename.ts [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+        ) {
+          message = codebridgeI18n.duplicateSupportFileError({fileName});
+        }
+      }
+
+      return message;
+    },
+    []
+  );
+};
```
Comment (thomasoniii) [inline apps/src/codebridge/FileBrowser/hooks/useCheckForDuplicateFilename.ts]: this and `checkForDuplicateFoldername` below were the existing functions, just yanked out into separate files for sake of tidiness.

File: apps/src/codebridge/FileBrowser/index.tsx [hunk 4]
```diff
@@ -762,13 +743,45 @@ export const FileBrowser = React.memo(() => {
...
+            project.files
+          );
+          if (duplicate) {
+            dialogControl?.showDialog({
+              type: DialogType.GenericAlert,
+              title: duplicate,
+            });
+          } else {
+            moveFile(e.active.data.current.id as string, e.over.id as string);
+          }
```
Comment (thomasoniii) [inline apps/src/codebridge/FileBrowser/index.tsx:737]: these duplicate checks should've gone into the prior DnD PR. I don't know what happened to that commit that included them, which is also the one that had the `DragType` enum originally.

File: apps/src/codebridge/Editor/index.tsx (hunk 1 above)
Comment (molly-moen) [reply apps/src/codebridge/Editor/index.tsx:49]: Do we want to allow image uploads? Do programs still run if there's an image file in the project?

File: apps/i18n/codebridge/en_us.json [hunk 5]
```diff
@@ -31,9 +31,9 @@
   "moveFolder": "Move folder",
   "moveFilePrompt": "Please enter your destination folder",
   "moveFolderPrompt": "Please enter your destination folder",
-  "newFile": "New File",
+  "newFile": "New file",
```
Comment (molly-moen) [inline apps/i18n/codebridge/en_us.json]: Why did you change the casing? The figma for the dropdown has both words in uppercase

File: apps/src/codebridge/FileBrowser/FileBrowserHeaderPopUpButton.tsx [hunk 6]
```diff
@@ -18,27 +23,59 @@ type FileBrowserHeaderPopUpButtonProps = {
...
+        className={classNames(
+          darkModeStyles.dropdownItem,
+          moduleStyles.dropdownItem
+        )}
+      >
+        <FontAwesomeV6Icon iconName="plus" iconStyle="solid" />
+        <div>{codebridgeI18n.newFile()}</div>
+      </div>
+      <FileUploader
+        callback={(fileName, contents) =>
```
Comment (molly-moen) [inline apps/src/codebridge/FileBrowser/FileBrowserHeaderPopUpButton.tsx:56]: nit: move these callbacks into helper methods? The logic seems to be repeated between here and the file dropdown, and I find it easier to read a method when it's not inline as a prop (that feedback is more for `errorCallback`, `callback` is just one line so 🤷‍♀️ )

File: apps/src/codebridge/FileBrowser/FileUploader.tsx [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+    const changeHandler = useCallback(() => {
+      const file = inputRef.current?.files?.[0];
+      if (file) {
+        if (!isValidMimeType(file.type)) {
+          sendCodebridgeAnalyticsEvent(
+            EVENTS.CODEBRIDGE_UPLOAD_UNACCEPTED_FILE,
+            appName,
+            {name: file.name, type: file.type}
+          );
+          errorCallback(`Cannot upload files of type ${file.type}`);
```
Comment (molly-moen) [inline apps/src/codebridge/FileBrowser/FileUploader.tsx]: We should make this string translatable

File: apps/src/codebridge/FileBrowser/hooks/useCheckForDuplicateFoldername.ts [hunk 8]
```diff
@@ -0,0 +1,27 @@
+import {ProjectFolder} from '@codebridge/types';
+import {useCallback} from 'react';
+
+import codebridgeI18n from '@cdo/apps/codebridge/locale';
+
+// Check if the foldername is already in use in the given folder.
+// If it is, alert the user and return true, otherwise return false.
+export const useCheckForDuplicateFoldername = () => {
```
Comment (molly-moen) [inline apps/src/codebridge/FileBrowser/hooks/useCheckForDuplicateFoldername.ts]: for my understanding: why is this a hook rather than a helper method?

File: apps/src/codebridge/Editor/index.tsx (hunk 1 above)
Comment (thomasoniii) [reply apps/src/codebridge/Editor/index.tsx:49]: Oh. Do we not? I must've been thinking of weblab. Locked it down so it's only text types, but will confirm with Sam et. al.

**Review by molly-moen — COMMENTED** [review]
Looks good, a couple small things!

File: apps/i18n/codebridge/en_us.json (hunk 5 above)
Comment (thomasoniii) [reply apps/i18n/codebridge/en_us.json]: Oh, maybe I changed it in the opposite direction.

But it's to make it consistent with the text in the pop up kebab menus, none of which have their second words capitalized. Which figma is the most recent version with the kebab menu? Casing should be consistent across the two pop ups.

File: apps/src/codebridge/FileBrowser/hooks/useCheckForDuplicateFoldername.ts (hunk 8 above)
Comment (thomasoniii) [reply apps/src/codebridge/FileBrowser/hooks/useCheckForDuplicateFoldername.ts]: It wasn't actually necessary - it was a historical artifact. I just made it into a helper function instead.

Looking through the blames, it looks like the function was originally added inline into `FileBrowser`, but this was triggering extra re-renders along the chain because it was newly created on each render.

Then those functions were retooled into memoized functions to prevent the re-renders, and then on this pass I just yanked those memoized values out into separate files w/o paying attention to them. But it wasn't memoizing any specific data, the memoize was just to prevent the re-renders when it was inline.

Anyway, moved around into functions instead. Good catch!

File: apps/i18n/codebridge/en_us.json (hunk 5 above)
Comment (molly-moen) [reply apps/i18n/codebridge/en_us.json]: This is the most up to date [figma](https://www.figma.com/design/vcNGOFsIks0HhRk6GEOfZf/Upper-Lab?node-id=4329-28214&node-type=canvas&m=dev). New file in the kebab menu isn't in the figma so I would say go with capitalizing all words in the menus based on what's there.

File: apps/i18n/codebridge/en_us.json [hunk 9]
```diff
@@ -8,27 +7,29 @@
...
-  "invalidNameError": "Names must be only letters, numbers, or dots",
+  "invalidFileUpload": "Cannot upload files of type {fileType}",
+  "invalidNameError": "Names must be only letters, numbers, or underscores",
   "loadingEnvironmentTooltip": "Your environment is loading. Please wait.",
   "makeStarter": "Make starter file",
   "makeSupport": "Make support file",
   "makeValidation": "Make validation file",
-  "moveFile": "Move file",
-  "moveFolder": "Move folder",
+  "moveFile": "Move",
```
Comment (molly-moen) [inline apps/i18n/codebridge/en_us.json:31]: should we consolidate these into one string? It will be consistently translated if we do that (same for delete and rename)

File: apps/src/codebridge/FileBrowser/hooks/useCheckForDuplicateFilename.ts [hunk 10]
```diff
@@ -0,0 +1,34 @@
+import {ProjectFile} from '@codebridge/types';
+import {useCallback} from 'react';
+
+import codebridgeI18n from '@cdo/apps/codebridge/locale';
+import {ProjectFileType} from '@cdo/apps/lab2/types';
+
+// Check if the filename is already in use in the given folder.
+// If it is, alert the user and return true, otherwise return false.
+export const useCheckForDuplicateFilename = () => {
```
Comment (molly-moen) [inline apps/src/codebridge/FileBrowser/hooks/useCheckForDuplicateFilename.ts]: should this be deleted now?

**Review by molly-moen — APPROVED** [review]
Couple cleanup comments, otherwise looks good!

File: apps/src/codebridge/FileBrowser/hooks/useCheckForDuplicateFilename.ts (hunk 10 above)
Comment (thomasoniii) [reply apps/src/codebridge/FileBrowser/hooks/useCheckForDuplicateFilename.ts]: I thought I had deleted it. Nuked.

File: apps/i18n/codebridge/en_us.json (hunk 9 above)
Comment (thomasoniii) [reply apps/i18n/codebridge/en_us.json:31]: I think we should keep them separate for now at least, but it'd probably be worth following up with product/design to confirm.

I like having them separate so it's trivial to change the display text for each individually, but if we're completely guaranteed not to do that, we should consolidate. I think it's a nice-to-have, after confirming.
